### ISSUE TITLE
database is locked on long running job

### ISSUE THREAD

**Issue by setop** [benbjohnson/litestream#218]
continuation of #58 

I have a daily job, written python (3.8).
By design, it starts after litestream, which is always running.
The job perform some insert and updates, generally 1000, at a pace of one to two per second.
Some day, everything is fine, but once every couple of days, the script outputs errors : `sqlite3.OperationalError: database is locked`. Last time, it appends mid time (after 450 inserts) and all the way to the end : meaning not a single update or insert has been successful after that very first error.

Is there a way to prevent that ?

**Comment by benbjohnson** [comment]
@setop What version of Litestream are you using? Litesteam used to obtain a write lock frequently but it's been reduced significantly by https://github.com/benbjohnson/litestream/pull/170 in v0.3.4.

Litestream still requires a write lock during WAL checkpointing so you'll need to set a `busy_timeout` to avoid the `database is locked` error: https://litestream.io/tips/#busy-timeout

**Comment by benbjohnson** [comment]
@setop Did you have any luck with setting the `busy_timeout` in your application?

**Comment by setop** [comment]
@benbjohnson, yes it is setup now. I'll let it run a couple of days and report the result here.
Thanks for the tip and thanks for the support.

**Comment by setop** [comment]
@benbjohnson, unfortunately, it happened again today after some days of good run.

Also, litestream was preventing truncation of journal. The journal was growing quite fast. With litestream disable I've been able to truncate it.

**Comment by benbjohnson** [comment]
@setop Hmm, that's odd. I have a server process that continually writes about 10 tx/sec all the time so 1 or 2 tx/sec shouldn't be a problem.

Is your application issuing a `PRAGMA wal_checkpoint()` command? Litestream maintains an open read transaction all the time to prevent checkpointing so that command will fail for `PASSIVE` checkpoints or hang for `TRUNCATE` or `RESTART` checkpoints.

**Comment by setop** [comment]
No `PRAGMA wal_checkpoint()` in my application. Only basic `select|insert|update`.

**Comment by benbjohnson** [comment]
@setop Can you provide some details about what Litestream version you're running, what operating system, and what CPU architecture? If you have a small script that can reproduce it that'd be great but I can try to repro on my side too.

**Comment by setop** [comment]
Hi,
Sorry for the delay. Due to this issue I removed Litestream from my system, replaced it with an old fashion backup, then time passed.
Since this project is very active and the it would be very valuable to me, I recently made a new attempt with version `v0.3.8`.
I also ensure that the pragmas are properly set in my application :
```
PRAGMA journal_mode = WAL
PRAGMA synchronous = NORMAL
PRAGMA busy_timeout = 5000
PRAGMA wal_autocheckpoint = 0
```

Unfortunately, nothing changed. I still have `sqlite3.OperationalError: database is locked`at some point and the process crash.

**Comment by ribtoks** [comment]
Same issue here. Using latest v0.3.9 as a systemd service and also having a simple Python script that from time to time does an insert. These inserts fail with `"database is locked"` error.

**Comment by fractaledmind** [comment]
Same issue here on v0.3.9. I am investigating to try and learn more, but because the database is locked rarely and randomly, investigation is difficult.

**Comment by loxs** [comment]
This seems to be a problem with Python's sqlite driver: https://forum.djangoproject.com/t/sqlite-and-database-is-locked-error/26994

I have many databases that I backup with litestream and only the one written in python has this problem. Tried changing many options as in the comments above, nothing short of stopping litestream helps my python app to be able to finish.

**Comment by hifi** [comment]
If you use transactions with your application always use BEGIN IMMEDIATE if Litestream is running and it should prevent the locked issue given you have busy timeout set.
